replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -13,140 +30,103 @@ private[journal] sealed abstract class Batch extends Product {
-                if (a.to > b.to) {
-                  val cleanTail: List[Batch] = tail.mapFilter {
-                    case appends: Appends => dropDeleted(appends, a)
-                    case delete: Delete   => delete.some
-                    case purge: Purge     => purge.some
-                  }
-
-                  val delete = deleteOf(a.to, b.origin orElse a.origin, b.version orElse a.version)
-                  delete :: cleanTail
-                } else {
-                  val delete = b.copy(offset = offset, origin = b.origin orElse a.origin, version = b.version orElse a.version)
-                  delete :: tail
-                }
-
-              case (_: Delete, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Purge, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: Nil
-
-              case (b: Purge, _: Action.Mark) =>
-                b.copy(offset = offset) :: Nil
-
-              case (b: Purge, a: Action.Delete) =>
-                deleteOf(a.to, a.origin, a.version) :: b :: Nil
-
-              case (_: Purge, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-            }
-
-          case Nil =>
-            record.action match {
-              case a: Action.Append => appendsOf(NonEmptyList.of(actionRecord(a))) :: Nil
-              case _: Action.Mark   => Nil
-              case a: Action.Delete => deleteOf(a.to, a.origin, a.version) :: Nil
-              case a: Action.Purge  => purgeOf(a.origin, a.version) :: Nil
-            }
+    def batches: List[Batch] = {
+      // we can drop first `append`, if `deleteTo` will discard it AND there is at least one more `append`
+      // we have to preserve one `append` to store latest `seqNr` and populate `expireAfter`
+      val appends = {
+        this.appends.flatMap { appends =>
+          val deleteTo = this.delete.map(_.to.value)
+          val records  = appends.records
+          val actions =
+            if (deleteTo.contains(records.head.action.range.to)) NonEmptyList.fromList(records.tail).getOrElse(records)
+            else records
+          appends.copy(records = actions).some
         }
       }
-      .foldLeft(List.empty[Batch]) { (bs, b) => // reverse order of batches
-        b match {
-          case b: Appends => b.copy(records = b.records.reverse) :: bs // reverse append actions
-          case b: Delete  => b :: bs
-          case b: Purge   => b :: bs
-        }
+
+      // if `delete` was not last action, adjust `delete`'s batch offset to update `metajournal` correctly
+      val delete = appends match {
+        case Some(appends) => this.delete.map(delete => delete.copy(offset = delete.offset max appends.offset))
+        case None          => this.delete
       }
+
+      // apply action batches in order: `Purge`, `Append`s and `Delete`
+      List(purge, appends, delete).flatten
```

Comment:
I run extra validation on our load-testing environment with code like:
```scala
      (delete, appends) match {
        case (Some(delete), Some(appends)) if delete.to.value > appends.records.last.action.range.to =>
          log.warn("unexpected situation: `delete.to > appends.seqNr`; falling back to previous batching version")
          // use version of `Batch` from current `master` branch
        case _ =>
          List(purge, appends, delete).flatten
      }
```

over 20+ hours there were no cases, when previous version was used - all our internal clients issue `delete` action with explicitly set `seqNr` which is aligned with aggregate's state (doesn't attempt to "delete in future")

Reply:
I just wanted to double check that `log.warn` does not return `IO[Unit]` in this case :thinking:

Reply:
no, it doesn't - there were more lines instead of ```// use version of `Batch` from current `master` branch``` - I just didn't want to post them all here 😄 

I just wanted to be sure, that we do not have active clients, which use `delete(max)`, though it is not expected to be used in "normal" case

it might be, that we have some weird "reset logic" where `delete(max)` makes sense and is used rarely - and this PR would break such journal, if `delete(max)` and next `append` would happen to land in same `poll`

using `delete(max)` makes sense only when journal is broken and somebody manually produces such action with hope that blank state would allow to blindly "reset" aggregate, while explicit `purge` could make more sense

Reply:
I am wondering, if there is a need for explicit handling in case of `delete(max)`?

Technically, if journal has `seqNr: 7`, then `delete(13)` would also act as `delete(max)` - thus special handling for `delete(max)` would be incomplete and `Batch` has no visibility over current state of journal and has no means to check if `delete(x)` would overshoot current `seqNr`.

In addition, `delete(x)` is legit way to [re]start journal from `seqNr: x` - thus one has to be very sure that journal exists in Cassandra before it can use `delete(max)` as otherwise one would create empty journal with no way to publish even one one additional event - `SeqNr.max + 1` would generate exception due to overflow.

It should be safe to merge this PR:
* there is no good reason to overshoot with `Delete` action in regular flow and if business logic tries to overshoot with `Delete`, then it means that it lacks consistency
* if one publishes `Delete` action manually, then one has breached single-writer principle and must be fully aware of internals of `kafka-journal` - thus should be able to act in informed way or be ready to do manual cleanup afterwards